### DISCUSSION TITLE
How does silero-vad compares to pyannote and nvidia nemo

### DISCUSSION THREAD

**Discussion by seekingdeep** [snakers4/silero-vad#152] [Q&A] [answered]
Hi there @snakers4 
how does silero-vad compares to pyannote and nvidia nemo MarbleNet?

https://github.com/pyannote/pyannote-audio
https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/Voice_Activity_Detection.ipynb

**Comment by snakers4** [comment, marked as answer]
Hi,

As for [pyannote](https://github.com/pyannote/pyannote-audio-hub#usage-1), we compared it previously, it was ok for whole audios, but not very fast. But am not sure that it is a out-of-the-box solution (it was trained on limited academic data) and I believe it does not support streaming (or we did not read all of their code to find the streaming examples). If you find some checkpoints and streaming examples, please send a link.

> Therefore, we use Google Speech Commands Dataset V2 [22] as speech data
> ![image](https://user-images.githubusercontent.com/12515440/148090471-4a5ce219-6bdf-4d38-b103-f9661d1b9e7f.png)

As for [MarbleNet](https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/Voice_Activity_Detection.ipynb) I found some streaming examples [here](https://github.com/NVIDIA/NeMo/blob/stable/tutorials/asr/Online_Offline_Microphone_VAD_Demo.ipynb) with pretrained models (but I am not sure if they were pre-trained just on Google speech commands which is limited).

In any case I am reluctant to invest time in adding this network into our benchmark for a couple of reasons:

- It was trained on limited data;
- It does not support real streaming (the 630ms window approach with 10ms stride seems a bit rough on the edges to me even compared to our first models with 100ms or 250ms windows, especially in production it would mean ~100x more excess computation, i.e. to get 60ms resolution, you need to run approximately 11 * 630ms worth of data compared to just 60ms);
- While the code in the notebook does not seem scary, they invested little time to make it easily separable from the Nemo toolkit;

**Reply by hbredin** [reply]
I wrote a blog post a while back describing how to use pyannote for streaming VAD:

https://herve.niderb.fr/fastpages/2021/08/05/Streaming-voice-activity-detection-with-pyannote.html

**Reply by snakers4** [reply]
> Each subsequent call to next(buffer) returns the current content of the 5s rolling buffer:
> For instance, if we are allowed 2s latency, we could benefit from the multiple overlapping buffers and combine them to get a better estimate of the speech probability in regions where the model is not quite confident (e.g. just before t=4s).

![image](https://user-images.githubusercontent.com/12515440/148496892-f1cd692d-7f87-47d2-8c86-90cfe7824dd3.png)
![image](https://user-images.githubusercontent.com/12515440/148496875-b5c7c662-62ea-444c-b046-e1004db88caa.png)
![image](https://user-images.githubusercontent.com/12515440/148496917-97227dee-de65-4c07-af7e-5b0ecd3ac9d3.png)

From the example above, looks like the model works on second long audio chunks.
Also I am not sure why erect all of the abstractions in the blog post, when just tensor or list slicing one-liner would do.

**Comment by seekingdeep** [comment]
Thanks for the detailed reply,

The pyannote link that you referenced links to pyannote 1,
since then a [paper](https://arxiv.org/abs/2104.04045) was released in 2021 revealing [pyannote 2.0 ](https://github.com/pyannote/pyannote-audio/tree/develop) which includes benchmarks against silero-vad.

Also, i am interested in hearing your thoughts about [SpeechBrain's vad](https://huggingface.co/speechbrain/vad-crdnn-libriparty)

**Reply by snakers4** [reply]
> since then a paper was released in 2021 revealing pyannote 2.0 which includes benchmarks against silero-vad.
>  It took around 3 days using 4 V100 GPUs to reach peak performance.

I believe I saw it that when the paper came out.
Most likely they also used a previous version of our VAD at that time and did not tune the hyper-params, which our old VAD required to do per domain.
I noted also that they used ludicrous compute levels for a VAD and overfit on small datasets.

Still, I followed the current link you provided, is there a streaming interface now?

The idea behind an old iteration of our VAD was to fit on a much larger corpus with a model working on kind of long chunks, but then we more drifted towards actually making our VAD usable in real life with little to no setup required (no GPUs, no large external libraries, no bloat, no complex setup or monolithic codebase,  no strings attached, no "peeking into the future", short adaptive chunks, etc).

> Also, i am interested in hearing your thoughts about SpeechBrain's vad

Just from reading the above page, a lot of words about GPUs, but I cannot see how the trained it. I probably can derive this by reading their code entirely, but I am kind of reluctant to invest time into this.

From the model name, probably this is FCN CNN fit on corpora similar to that of Nvidia's solution.

_____________________________

In any case, the only way to properly answer your questions would be to invest time and effort into setting up each of the above VADs in a separate environment, fix the conflicts and then apply them in a kind of similar [fashion](https://github.com/snakers4/silero-vad/wiki/Quality-Metrics#method), but keeping in mind that they would "cheat" by looking into the future or we would use 100x more compute by applying their VADs on overlapping audio chunks, from which we so much tried to depart from.

It would be an interesting exercise, but is it worth the effort? 

These solutions clearly have a different set of values and target criteria (i.e. achieve max quality on a given limited dataset regardless of the practicality as opposed to providing a working simple general solution), though to their credit, their models at least are not huge.

**Comment by seekingdeep** [comment]
Thank you for your detailed reply,

Note that SpeechBrain still [don't](https://github.com/speechbrain/speechbrain/issues/1205) have stream inferencing, nor do i see a deployment pipeline, but one of the things that caught my eye is that their model was able to correctly predict the part where both music and speech are active and classify it as speech [in this sample.](https://www.dropbox.com/s/lkngxlybrh4zamh/example_vad_music.wav)
I am intrigued in finding silero vad performance on the same sample

For speechbrain here are the links so you can test and compare:

- VAD inferencing pipeline:
https://colab.research.google.com/drive/1Msk2cgSEw-jCuXHmz2_-iOrv-gJHCCxu?usp=sharing

- SpeechBrain released Vad model download:
https://drive.google.com/drive/folders/1YLYGuiyuTH0D7fXOOp6cMddfQoM74o-Y?usp=sharing

**Reply by snakers4** [reply]
> that their model was able to correctly predict the part where both music and speech are active and classify it as speech in this sample.
I am intrigued in finding silero vad performance on the same sample

You can do the same in the notebook we [provide](https://colab.research.google.com/github/snakers4/silero-vad/blob/master/silero-vad.ipynb) in 2 or 3 clicks:

![image](https://user-images.githubusercontent.com/12515440/148498184-1308660e-e9de-472f-aba6-18c3e33ab10f.png)

Github does not host audios yet, but if you listen to `only_speech.wav` that the notebook creates out of the box, it works perfectly on this audio file. You can fiddle a bit with hyper-params to get some space around the last syllable, but this is fine out-of-the box as well.

**Comment by seekingdeep** [comment]
silero's performance is actually great, you can note that adding an auto-postprocessing step to the pipeline inorder to automatically generate the time per sentence would be even better.
your work is appreciated, wishing you good luck

**Reply by snakers4** [reply]
> that adding an auto-postprocessing step to the pipeline 

The whole model usage is so minimalist, that everything besides these lines is more or less sugar or exists for convenience:

https://github.com/snakers4/silero-vad/blob/76687cbe25ffdf992ad824a36bfe73f6ae1afe72/utils_vad.py#L206-L211

Basically you can do whatever with these probs - just plot them and decide.
We actually have post-processing in [streaming](https://github.com/snakers4/silero-vad/blob/master/utils_vad.py#L406-L425) and [non-streaming](https://github.com/snakers4/silero-vad/blob/master/utils_vad.py#L213-L240) examples.

> in order to automatically generate the time per sentence would be even better.

If I understand you correctly, the `return_seconds` is what you are looking for:

https://github.com/snakers4/silero-vad/blob/76687cbe25ffdf992ad824a36bfe73f6ae1afe72/utils_vad.py#L119-L128

**Reply by seekingdeep** [reply]
i was suggesting:
- Auto-hyperparameter tuning
- Auto-postprocessing

**Reply by snakers4** [reply]
> Auto-hyperparameter tuning

We mostly got rid of the necessity to really tune the params with new release.
If you are seriously using the VAD in some real life scenario, 5 minutes to build a chart and change the prob / silence durations is not a big deal.

> Auto-postprocessing

With the examples we provide, post-processing is done automatically.
We provide the most simplistic model API possible, i.e. `speech_prob = model(chunk, sampling_rate)`.
We could of course hide the logic inside of JIT containers, but most likely that would cause more problems because we cannot know all cases in advance.

**Comment by seekingdeep** [comment]
> 5 minutes to build a chart and change the prob / silence durations

can you explain.

**Reply by snakers4** [reply]
typically most important VAD hyper-params (i.e. threshold) are tuned once per domain, which takes several minutes, provided the data in your domain is consistent
